### PR TITLE
Several patches to improve the quality of suggestions

### PR DESCRIPTION
**Pull request by tcwanggoogle**
This CL contains all the improvement I made on the source code after Suggestion branch. The important improvement are included in the following commits:
(1) Add support of bitcast on GEP (6878ca6)
(2) Add support to detect struct array and relative unit test (bd4ca56) and Fix a bug to handle directly GEP on struct array (f57914d)
(3) Add support to analyze function args with one field (36c4021)
(4) Add support to analyze multiple fields passed into same argument (c000a13)
And other commits that fix the statistics counter and minor issues.

### CODE REVIEW

**Comment by tcwanggoogle** [comment]
I found I haven't added detailed header comments for the functions in StructFieldAccessInfo.cpp. I'll add them later.

**Comment by tcwanggoogle** [comment]
Comments added.

File: lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h [hunk 1]
```diff
@@ -45,9 +45,12 @@ using namespace llvm;
 
 namespace llvm {
 typedef unsigned FieldNumType;
+typedef unsigned ArgNumType;
 typedef std::pair<const StructType *, FieldNumType> StructInfoMapPairType;
 typedef uint64_t ProfileCountType;
-typedef std::vector<FieldNumType> FieldNumArrayType;
+typedef std::pair<FieldNumType, ProfileCountType> FieldAccessInfoPairType;
+typedef std::vector<FieldAccessInfoPairType> ArgFieldMappingArrayType;
+typedef std::vector<ArgFieldMappingArrayType *> ArgFieldMappingArrayArrayType;
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldCacheAnalysisImpl.h:53]: I understand you want to use typedef, but frankly the type name is too long and not very meaningful.

File: lib/Transforms/IPO/StructAnalysisCloseProximity.cpp [hunk 2]
```diff
@@ -385,39 +443,29 @@ CloseProximityBuilder::buildFieldReferenceGraph(const Function *F) {
...
-                     << " and data " << D << " bytes \n");
-          BBI->LastNode = NewNode;
-        } else {
-          // If there's no previous node, update BBI info
-          DEBUG_WITH_TYPE(
-              DEBUG_TYPE_FRG,
-              dbgs() << "No previous node found. It is the first node\n");
-          BBI->FirstNode = BBI->LastNode = NewNode;
-        }
+        createNewFRGNode(FRG, BBI, NewNode, &I);
```
Comment (eramandev) [inline lib/Transforms/IPO/StructAnalysisCloseProximity.cpp]: Other than the fact that you can't use auto, it probably makes sense to move the call to lambda outside the if then else. Also, perhaps change the name of the lambda. It looks weird to call FRG->createNode and then call createNewFRGNode

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp [hunk 3]
```diff
@@ -29,15 +29,27 @@
...
-    "struct-analysis-number-buckets", cl::init(10), cl::Hidden,
-    cl::desc("Number of buckets used to analyze struct hotness"));
+static cl::opt<unsigned>
+    HistogramSizeForStats("struct-analysis-number-buckets", cl::init(10),
+                          cl::Hidden,
+                          cl::desc("Number of buckets used in the histogram "
+                                   "for analyzing struct hotness"));
+
+static cl::opt<bool> FilterOutZeroHotness(
+    "struct-analysis-filter-zero-hotness", cl::init(false), cl::Hidden,
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldAccessInfo.cpp]: True is perhaps a better default here.

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp [hunk 4]
```diff
@@ -60,24 +72,75 @@ void StructFieldAccessInfo::addFieldAccessNum(const Instruction *I,
...
+// Return the field number if the argument has a mapping to a field and
+// the hotness is non-zero. Otherwise, return None type of Optional
+Optional<FieldNumType> StructFieldAccessInfo::getUniqueArgFieldMapping(
+    FunctionCallInfoSummary *FuncSummary, ArgNumType ArgNum) const {
+  if (FuncSummary->AllArgFieldMappings.size() > 1)
+    return None;
+  auto *ArgFieldMapping = FuncSummary->AllArgFieldMappings[0];
+  assert(ArgNum < ArgFieldMapping->size());
+  auto &FieldAccessPair = (*ArgFieldMapping)[ArgNum];
+  // Only check the hotness info if FieldNum of this ArgNum is none-zero
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldAccessInfo.cpp]: typo: none-zero -> non-zero

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp [hunk 5]
```diff
@@ -86,28 +149,63 @@ StructFieldAccessInfo::calculateFieldNumFromGEP(const User *U) const {
...
-  auto *Index = cast<Constant>(Op);
-  auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
-  assert(Offset < NumElements);
-  // TODO: ignore indices after this one. If there's indices, the field has to
-  // be an array or struct
-  return Offset + 1; // return field number starting from 1
+  if (Op->getType()->getPointerElementType()->isStructTy() &&
+      Op->getType()->getPointerElementType() == StructureType) {
+    if (U->getNumOperands() <
+        3) // GEP to calculate struct field needs at least 2
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldAccessInfo.cpp]: Try to run clang-format before merging.

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp [hunk 6]
```diff
@@ -86,28 +149,63 @@ StructFieldAccessInfo::calculateFieldNumFromGEP(const User *U) const {
...
+    // Operand 2 should be the index to the field, and be a constant
+    Op = U->getOperand(2);
+    assert(isa<Constant>(Op));
+    auto *Index = cast<Constant>(Op);
+    auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
+    assert(Offset < NumElements);
+    // TODO: ignore indices after this one. If there's indices, the field has to
+    // be an array or struct
+    return Offset + 1; // return field number starting from 1
+  } else if (Op->getType()->getPointerElementType()->isArrayTy() &&
```
Comment (eramandev) [inline lib/Transforms/IPO/StructFieldAccessInfo.cpp]: This makes it hard to read. If you have a really complex condition in if, it is better to move it into a separate static function. May be add a FIXME here.

**Review by eramandev — APPROVED** [review]
I haven't thoroughly reviewed this pull request but I did go through most of this. Go ahead and merge this.

File: lib/Transforms/IPO/StructAnalysisCloseProximity.cpp (hunk 2 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructAnalysisCloseProximity.cpp]: I'll change the lambda name.

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp (hunk 3 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldAccessInfo.cpp]: Sure. That makes sense.

Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldAccessInfo.cpp]: But that involves changing a few unit tests that doesn't profiling. I can add flags in the unit tests.

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp (hunk 4 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldAccessInfo.cpp]: Thanks

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp (hunk 5 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldAccessInfo.cpp]: I think I did every time. I'll make sure I do it before next commit.

File: lib/Transforms/IPO/StructFieldAccessInfo.cpp (hunk 6 above)
Comment (tcwanggoogle) [reply lib/Transforms/IPO/StructFieldAccessInfo.cpp]: I'll create a static function

**Pull request merged**